Boot-to-ramdisk image generation scripts

boricj opened the pull request:
Using the ramdisk generation functions I've developed for PXE booting in #94, I managed to boot MINIX directly from a USB stick.

As with PXE booting, it uses the bootloader to load a big ramdisk into memory ; the USB disk isn't used after booting due to the lack of USB support on x86.

It's hacky, slow to boot and takes a lot of RAM, but it works. Tested on QEMU and on a Dell E6520. 

Note that booting on QEMU using the NetBSD bootloader takes about an hour on my computer. Also, while bundling tarball sets for setup could be possible, x86_usbimage.sh won't do it in order to keep boot time and memory requirements down to a reasonable level.

dcvmoole commented:
This is cool stuff. My only concern is that for PR reasons, we have to make absolutely sure that this cannot be interpreted as (x86) "USB support" or "USB boot support" or anything like that. Otherwise people will take this the wrong way, expect full USB support, be thoroughly disappointed, and complain to us (like before) (sigh). As such I'm kind of wondering: is this patch specific to USB at all? I mean, it would work for any situation where the bootloader is able to load the ramdisk into memory, not merely USB, right? Should it not be called "boot-to-ramdisk support" instead or so? @boricj what do you think? Not that this should get in the way of merging the patch by the way.

boricj commented:
The boot-to-ramdisk part itself isn't specific to USB, I lifted it verbatim from my PXE pull request. As long as you can _somehow_ get everything into memory (PXE, QEMU multiboot...) MINIX should boot.

As for the MBR-partitioned image which loads and runs a big ramdisk into memory, besides the obvious USB stick application I guess it could be used to boot from any block device a given BIOS can handle, such as SD cards for example.

I'll split the boot-to-ramdisk part from the disk image part. Also, embedding tarball sets for setup should be controlled with a single, common command-line option for all scripts. Finally, if SYSLINUX is not to be promoted to the same status as U-Boot, #94 may nevertheless be closed once step-by-step instructions for PXE booting are written on the wiki, at the expense of the one-liner script.

If PR is a problem, then I propose to call this the _Universally Supported Boot disk image_, the image to try if anything else fails :-)

boricj commented:
I made the modifications and also did some clean-up of the image generation framework.

I found an odd issue : if I try to perform an installation with this branch on QEMU, I get `fs_cache: I/O error on device 3/129, block 0 (0)`, which crashes the MFS instance of the hard drive with a stack trace and hangs the setup script (but not the OS). I also get a `VFS: path lookup error; root node not found` error after a while.

This is not limited to the installation CD : the boot-to-ramdisk image also exhibits the same behavior. I have no idea what's going on...

dcvmoole commented:
Hmm, device 3/129 is /dev/c0d0p0s1, which is typically the /home partition. I guess this happens when setup is trying to find out whether a reinstallation (ie preservation of /home) is possible? How it gets this far but then fails, I'm not sure. I thought MFS failed cleanly if eg the superblock isn't readable (it is reading block 0 here after all), but maybe something broke there? Maybe it's a good idea to make a separate issue out of this, because I doubt that this could actually be *caused* by a bug in this pull request and so it shouldn't keep us from merging it.

Speaking of which, I see no reason not to merge this patch, but I'm also not an expert when it comes to these scripts - so, unless you want to do further work on it or someone else speaks up soon, I'll just go ahead and pull it.

sambuc commented on releasetools/x86_hdimage.sh:
```diff
@@ -24,6 +20,10 @@ then
+# set up disk creation environment
+. releasetools/image.defaults
+. releasetools/image.functions
+
```
Did you check that moving those above the HDD default size doesn't break it. I mean, if image.defaults sets {HOME,USR,…}_SIZE, the commands above will never be executed.

sambuc commented on releasetools/x86_ramimage.sh:
```diff
@@ -0,0 +1,57 @@
+fi
+
+#: ${RAMDISK_SIZE=$(( 200*(2**20) ))}
+
```
Just like here, you first set defaults override, then source ```image.default```.

sambuc commented:
Other that my inline comments, it looks fine to me.

boricj commented:
I moved the defaults back where they should be for x86_hdimage.sh. I haven't realized I moved them while refactoring.

Otherwise I think I'm done, you can submit it to Gerrit.

I'll open a issue for the strange crash on setup once this is merged.

sambuc commented:
Submitted – twice oO – to gerrit. Let's see what jenkins says.

sambuc commented:
Merged in a7a79fa .
